diff --git a/dist/build/templates/app-page-runtime.js b/dist/build/templates/app-page-runtime.js
index f1ab03f292340fe08b563990110ac0835d3415fb..a3e659f5558b609ab4ec7e82726a0c07685ed279 100644
--- a/dist/build/templates/app-page-runtime.js
+++ b/dist/build/templates/app-page-runtime.js
@@ -1,4 +1,6 @@
 "use strict";
+// Lumis: Next.js #85999 overwrites application Vary fields. Use jshttp/vary until upstream fixes it.
+const vary = require("vary");
 Object.defineProperty(exports, "__esModule", {
     value: true
 });
@@ -456,7 +458,7 @@ function createAppPageEntrypoint({ tree, page, pathname, require: __next_app_req
         };
         try {
             const varyHeader = routeModule.getVaryHeader(resolvedPathname, interceptionRoutePatterns);
-            res.setHeader('Vary', varyHeader);
+            vary(res, varyHeader);
             let parentSpan;
             const invokeRouteModule = async (span, context)=>{
                 const nextReq = new _node.NodeNextRequest(req);
diff --git a/dist/esm/build/templates/app-page-runtime.js b/dist/esm/build/templates/app-page-runtime.js
index 1970588724b6e4880f27a8be60ca32ae6437bc28..5735b99e829047fec179f791b90ebdafb5e29524 100644
--- a/dist/esm/build/templates/app-page-runtime.js
+++ b/dist/esm/build/templates/app-page-runtime.js
@@ -1,3 +1,5 @@
+// Lumis: Next.js #85999 overwrites application Vary fields. Use jshttp/vary until upstream fixes it.
+import vary from 'vary';
 import { AppPageRouteModule } from '../../server/route-modules/app-page/module.compiled' with {
     'turbopack-transition': 'next-ssr'
 };
@@ -471,7 +473,7 @@ export function createAppPageEntrypoint({ tree, page, pathname, require: __next_
         };
         try {
             const varyHeader = routeModule.getVaryHeader(resolvedPathname, interceptionRoutePatterns);
-            res.setHeader('Vary', varyHeader);
+            vary(res, varyHeader);
             let parentSpan;
             const invokeRouteModule = async (span, context)=>{
                 const nextReq = new NodeNextRequest(req);
